source/common/config/grpc_subscription_impl.cc

```diff
@@ -63,14 +65,37 @@ void GrpcSubscriptionImpl::onConfigUpdate(const std::vector<Config::DecodedResou
   // supply those versions to onConfigUpdate() along with the xDS response ("system")
   // version_info. This way, both types of versions can be tracked and exposed for debugging by
   // the configuration update targets.
+  auto start = dispatcher_.timeSource().monotonicTime();
   callbacks_.onConfigUpdate(resources, version_info);
+  std::chrono::milliseconds update_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+      dispatcher_.timeSource().monotonicTime() - start);
   stats_.update_success_.inc();
   stats_.update_attempt_.inc();
   stats_.update_time_.set(DateUtil::nowToMilliseconds(dispatcher_.timeSource()));
   stats_.version_.set(HashUtil::xxHash64(version_info));
   stats_.version_text_.set(version_info);
+  stats_.update_duration_.recordValue(update_duration.count());
   ENVOY_LOG(debug, "gRPC config for {} accepted with {} resources with version {}", type_url_,
             resources.size(), version_info);
+
+  if (update_duration > updateDurationLogThreshold()) {
+    // Figure out the total size of all resource names plus delimiters
+    int name_size = 0;
+    std::for_each(resources.begin(), resources.end(),
+                  [&name_size](const Config::DecodedResource& resource) {
+                    name_size += resource.name().size() + 2;
+                  });
+    std::string resource_names;
+    resource_names.reserve(name_size);
+    for (const auto& resource : resources) {
+      resource_names.append(resource.get().name());
+      if (&resource != &resources.back()) {
+        resource_names.append(", ");
+      }
+    }
```

Comment:
use StrJoin with join-formatters https://abseil.io/docs/cpp/guides/strings#join-formatters and it should be in the line of ENVOY_LOG so it's not evaluated when the log isn't desired.

Reply:
Good point, and done.